docs/export-queue.md

```diff
@@ -0,0 +1,214 @@
+        
+        //add a new entry to the inverted index
+        querySystem.insertRow(newRow, row);
+        //remove the old entry from the inverted index
+        querySystem.deleteRow(oldRow, row);
+    }
+}
+```
+
+The above example would keep the external index up to date beautifully as long
+as the following conditions are met.
+
+  * Threads executing transaction always complete.
+  * Only a single thread ever responds to a notification.
+
+However these conditions are not guaranteed by Fluo.  Multiple threads may
+attempt to process a notification concurrently (only one may succeed).  Also at
+any point in time a transaction may fail (for example the computer executing it
+may reboot).   Both of these problems will occur and will lead to corruption of
+the external index in the example.  The inverted index and Fluo  will become
+inconsistent.  The inverted index will end up with multiple entries (that are
+never cleaned up) for single entity even though the intent is to only have one.
+
+The root of the problem in the example above is that its exporting uncommitted
+data.  There is no guarantee that setting the column `<row>:meta:counter1` to
+`newCount` will succeed until the transaction is successfully committed.
+However, `newRow` is derived from `newCount` and written to the external query
+system before the transaction is committed (Note : for observers, the
+transaction is committed by the framework after `process(...)` is called).  So
+if the transaction fails, the next time it runs it could compute a completely
+different value for `newRow` (and it would not delete what was written by the
+failed transaction).  
+
+## Solution 
+
+The simple solution to the problem of exporting uncommitted data is to only
+export committed data.  There are multiple ways to accomplish this.  This
+recipe offers a reusable implementation of one method.  This recipe has the
+following elements:
+
+ * An export queue that transactions can add key/values to.  Only if the transaction commits successfully will the key/value end up in the queue.  A Fluo applications can have multiple export queues, each one must have a unique id.
+ * When a key/value is added to the export queue, its given a sequence number.  This sequence number is guaranteed to increase.
+ * Each export queue is configured with an observer that processes key/values that were successfully committed to the queue.
+ * When key/values in an export queue are processed its deleted, so the export queue does not keep any long term data.
+ * Key/values in an export queue are placed in buckets.  This is done so that all of the updates in a bucket can be processed in a single transaction.  This allows an efficient implementation of this recipe in Fluo.  It can also lead to efficiency in a system being exported to, if the system can benefit from batching updates.  The number of buckets in an export queue is configurable.
+
+There are three steps to using this recipe :
```

Comment:
I like that.   I did not like "steps" when I wrote it, but I could not think of anything I liked better.

fixed in  5191b6f